### DISCUSSION TITLE
Can unnecessary nesting from former supertables be removed after Craft5 conversion?

### DISCUSSION THREAD

**Discussion by EFWall** [craftcms/cms#16275] [General]
Hello. We have recently converted our project to Craft5. The project has been around since early Craft3. It relied heavily on nested Matrix/Supertable combinations to create flexible content structures. Now, post conversion, we have several instances of unnecessary nested levels that I am wishing we could flatten. I am wondering if there is a (somewhat simple) way to accomplish this, or if the best practice is just to live with the unnecessary complexity, as it is not really a problem outside of being overly complex.

My thinking is that an ideal scenario would be the ability to move entries (in bulk) between matrix fields that share the same entry type, much like one can now do for entries between sections. If this were possible, it would be fairly easy for us to set up new matrix fields where we want them to be and just move all the entries over.

Is this possible? Or is there some other method to accomplish this? Or, as I suggested previously, should I just live with it as is?

An example structure for what I'm talking about would be something like this:

— EntryTypeA
— — MatrixFieldB / EntryTypeB (former static Supertable that only contained one block and one matrix field)
— — — MatrixFieldC (former matrix field containing several blocks, now several entry types, as below)
— — — — EntryTypeC 
— — — — — Entries we want to keep
— — — — EntryTypeD
— — — — — Entries we want to keep
— — — — EntryTypeE
— — — — — Entries we want to keep
— — — — EntryTypeF
— — — — — Entries we want to keep

And we'd be looking to make it like this (essentially eliminating the no longer necessary MatrixFieldB and EntryTypeB):

— EntryTypeA
— — MatrixFieldC (former matrix field containing several blocks, now several entry types, as below)
— — — EntryTypeC 
— — — — Entries we want to keep
— — — EntryTypeD
— — — — Entries we want to keep
— — — EntryTypeE
— — — — Entries we want to keep
— — — EntryTypeF
— — — — Entries we want to keep

**Comment by jannisborgers** [comment]
TLDR; Doing this with a content migration can save time.

While this is a couple of months old and I think it’s actually a problem Verbb would have to solve with their Craft 5 migration (if possible), I run into this often enough to document my solution here. It’s a rare situation where the templates might break, because while Super Table fields are converted to Matrix, if they were static, the way you access sub fields changes.

## Authoring in the control panel
Depending on the content structure and the fields within your Super Table fields, the AX gets bad quickly. Moving a couple of entries by hand between blocks would probably be doable – if it were possible.

This is the control panel UI from our current upgrade. Note that in our case, the static field is itself inside a content-builder matrix block, so there is another level of nesting here. The table shows the progress from Craft 4 to Craft 5, and a clean version that works like the Super table version, but native:

| Craft 4 | Craft 5 | Craft 5 (clean) |
| --------- | --------- | --------- |
| <img width="819" alt="craft5-static-super-table-matrix-nesting-01" src="https://github.com/user-attachments/assets/c40a8883-f65f-46d8-8a36-fccb973a0ffc" /> | <img width="812" alt="craft5-static-super-table-matrix-nesting-02" src="https://github.com/user-attachments/assets/f8a71e84-d14e-4fec-a672-a9caed229ae4" /> | <img width="814" alt="craft5-static-super-table-matrix-nesting-03" src="https://github.com/user-attachments/assets/9843dfe7-dd09-4365-a759-bc159ed46312" /> |
| `teasers` is the static Super Table field, `teaser_types` (handle not shown) is a nested matrix field offering different entry types. | `teasers` is now also a matrix field with repeating entries. An additional "Add entry" button is shown, and the template has to change (see above). | `teasers` holds the entry types previously owned by the nested matrix field `teaser_types` directly. `teaser_types` (both the entry type and the matrix field) can be removed. |

## Using a content migration
A content migration can to this in one go, barely an inconvenience:
* Preparation: 
  * Add the entry types available in the nested matrix field in the former static Super Table field to the former static Super Table field itself (e.g. add `EntryTypeC`, `EntryTypeD`, `EntryTypeE`, `EntryTypeF` to `MatrixFieldB`’s allowed entry types)
* Run the migration:
  * It finds all uses of the former static Super Table entries (`EntryTypeB`)
  * For each of the entries, it loops through the nested matrix field (`MatrixFieldC`)
  * For each entry in the nested matrix field, it creates a new entry, copies the values from the old entry, but sets the primaryOwnerID to the former static Super Table field `EntryTypeB`, essentially copying the entry to one level higher.
  * After each former Super Table entries has been processed, delete the original entries in the nested matrix field so that there is not duplicate content (clear `MatrixFieldC`).
* After the migration has run, do some manual clean-up: 
  * Check if the content has correctly changed position from `MatrixFieldC` to `MatrixFieldB`.
  * If successful, `EntryTypeB` and `MatrixFieldC` can both be removed.

Please be aware that this migration cannot be reverted, so test locally. Re-import the live DB if an attempt fails. I encountered some errors when the migration tried to save the new entries, this was due to foreign key constraint issues. I had to run `craft gc` in production to fix this.

Re-use with care, no guarantees. 

I have highlighted the lines where you can change the entry type `EntryTypeB` and the handle of `MatrixFieldC`:
```diff
<?php

namespace craft\contentmigrations;

use Craft;
use craft\db\Migration;
use craft\elements\Entry;

/**
* m250222_134936_flatten_former_static_super_table_field migration.
*/
class m250222_134936_flatten_former_static_super_table_field extends Migration
{
+    // Run as dry run to only log results
+    protected $dryRun = false;
+    protected $superTableEntryTypeHandle = 'EntryTypeB';
+    protected $nestedMatrixFieldHandle = 'MatrixFieldC';

    public function safeUp()
    {
        // Get all entries of the former static Super Table field’s entry type
        $superTableEntries = Entry::find()->type($this->superTableEntryTypeHandle)->all();
        Craft::info("Found " . count($superTableEntries) . " entries in former super table static field.");

        foreach ($superTableEntries as $index => $superTableEntry) {
            Craft::info("Processing entry " . $index . "/" . count($superTableEntries) . " of type '" . $superTableEntry->type->handle . "' with ID: " . $superTableEntry->id);

            // Process the entries inside the nested matrix field in the former Super Table static field/entry type
            foreach ($superTableEntry[$this->nestedMatrixFieldHandle] as $nestedMatrixEntry) {
                $this->processEntry($nestedMatrixEntry, $superTableEntry);
            }

            // After processing the entries, the originals within the nested matrix field can be deleted
            if (!Craft::$app->elements->deleteElement($superTableEntry)) {
                echo "Failed to delete old entry in former super table static field.";
            }
        }

        return true;
    }

    protected function processEntry($sourceEntry, $parentEntry)
    {
        Craft::info("Processing entry of type '" . $sourceEntry->type->handle . "' with ID: " . $sourceEntry->id);
        Craft::info($sourceEntry);
        if ($this->dryRun) {
            Craft::info("Field values:");
            Craft::info($sourceEntry->getFieldValues());
        } else {
            // Create new entry and make it a child of the former Super table field directly, instead of the nested matrix field
            $newEntry = new Entry();
            $newEntry->typeId = $sourceEntry->type->id;
            $newEntry->fieldId = $parentEntry->fieldId;
            $newEntry->primaryOwnerId = $parentEntry->getPrimaryOwnerId();
            // Copy all field values from the source entry to the new entry
            $newEntry->setFieldValues($sourceEntry->getFieldValues());

            // Save new entry
            if (!Craft::$app->elements->saveElement($newEntry)) {
                echo "Failed to save new entry based on source entry with ID: " . $sourceEntry->id;
            }
        }
    }

    public function safeDown()
    {
        echo "FlattenFormerStaticSuperTableField migration cannot be reverted.\n";
        return false;
    }
}
```
